src/core/runtime/dns-browser.js

```diff
@@ -1,33 +1,62 @@
-/* global self */
+/* eslint-env browser */
 'use strict'
 
-module.exports = (domain, opts, callback) => {
+const TLRU = require('../../utils/tlru')
+const { default: PQueue } = require('p-queue')
+const { default: ky } = require('ky-universal')
+const nodeify = require('promise-nodeify')
+
+// Avoid sending multiple queries for the same hostname by caching results
+const cache = new TLRU(1000)
+// TODO: /api/v0/dns does not return TTL yet: https://github.com/ipfs/go-ipfs/issues/5884
+// However we know browsers themselves cache DNS records for at least 1 minute,
+// which acts a provisional default ttl: https://stackoverflow.com/a/36917902/11518426
+const ttl = 60 * 1000
+
+// browsers limit concurrent connections per host,
+// we don't want preload calls to exhaust the limit (~6)
+const _httpQueue = new PQueue({ concurrency: 4 })
+
+// Delegated HTTP resolver sending DNSLink queries to ipfs.io
+// TODO: replace hardcoded host with configurable DNS over HTTPS: https://github.com/ipfs/js-ipfs/issues/2212
+const api = ky.create({
+  prefixUrl: 'https://ipfs.io/api/v0/',
+  hooks: {
+    afterResponse: [
+      async (input, options, response) => {
+        const query = new URL(response.url).search.slice(1)
+        const json = await response.json()
+        cache.set(query, json, ttl)
+      }
+    ]
+  }
+})
+
+const ipfsPath = (response) => {
+  if (response.Path) return response.Path
+  throw new Error(response.Message)
+}
+
+module.exports = (fqdn, opts = {}, cb) => {
```

Comment:
Can this be promise-first please?  E.g. make this an async method but export a nodeified version.  This will make future refactors to remove callbacks easier.

Reply:
I think its already done? the async method is inside: 
```js
const resolveDnslink = async (fqdn, opts = {}) => {
  ...
}
return nodeify(resolveDnslink(fqdn, opts), cb)
```

when we refactor to remove callbacks we just remove the wrapper.
Let me know if this should be handled some other way.